extensions/HyperVExtension/src/HyperVExtension/Models/VirtualMachineCreation/FileDownloadMonitor.cs

```diff
@@ -0,0 +1,107 @@
+/// download. As new progress comes in the FileDownloadMonitor publishes this progress to
+/// its subscribers.
+/// </summary>
+internal sealed class FileDownloadMonitor
+{
+    private readonly ILogger _log = Log.ForContext("SourceContext", nameof(FileDownloadMonitor));
+
+    private readonly object _lock = new();
+
+    private readonly List<IProgress<IOperationReport>> _subscriberList = new();
+
+    private readonly Progress<ByteTransferProgress> _progressReporter;
+
+    private bool _downloadInProgress;
+
+    private DownloadOperationReport? _lastSentReport;
+
+    public FileDownloadMonitor(IProgress<IOperationReport> progressProvider)
+    {
+        AddSubscriber(progressProvider);
+        _progressReporter = new(PublishProgress);
+    }
+
+    public void AddSubscriber(IProgress<IOperationReport> progressProvider)
+    {
+        lock (_lock)
+        {
+            if (_lastSentReport != null)
+            {
+                // Subscriber addition requested so we'll submit the last recorded
+                // progress we received before adding it to the subscriber list
+                progressProvider.Report(_lastSentReport);
+            }
+
+            _subscriberList.Add(progressProvider);
+        }
+    }
+
+    private void PublishProgress(ByteTransferProgress transferProgress)
+    {
+        _lastSentReport = new DownloadOperationReport(transferProgress);
+        _subscriberList.ForEach(subscriber => subscriber.Report(_lastSentReport));
+    }
+
+    public async Task StartAsync(
+        Stream source,
+        Stream destination,
+        int bufferSize,
+        long totalBytesToExtract,
+        CancellationToken cancellationToken)
+    {
+        lock (_lock)
+        {
+            if (_downloadInProgress)
+            {
+                // Download already started, no need to attempt
+                // to start it again.
+                return;
+            }
+
+            _downloadInProgress = true;
+
+            // Before starting the initial download, we'll update _lastSentReport so any
+            // subscribers who are added before and after we leave the lock are notified.
+            PublishProgress(new(bytesReceived: 0, totalBytesToReceive: totalBytesToExtract));
+        }
+
+        await source.CopyToAsync(
+            destination,
+            _progressReporter,
+            bufferSize,
+            totalBytesToExtract,
+            cancellationToken);
+
+        StopMonitor();
+    }
+
+    public void StopMonitor(string? errorMessage = null)
+    {
+        lock (_lock)
+        {
+            // Download already stopped.
+            if (!_downloadInProgress)
+            {
+                return;
+            }
+
+            _downloadInProgress = false;
+
+            // Send error message to all subscribers.
+            if (!string.IsNullOrEmpty(errorMessage))
+            {
+                PublishProgress(new ByteTransferProgress(errorMessage));
+            }

```

Comment:
Only error case here? No notification in case of successful completion?

Reply:
the successful notification happens within ByteTransferProgress itself. With 

`public bool CompletedSuccessfully => BytesReceived == TotalBytesToReceive;`

Before starting the [transfer](https://github.com/microsoft/devhome/blob/23f42ed9264be12ac66ce13d4bb82be535d5ee3a/extensions/HyperVExtension/src/HyperVExtension/Extensions/StreamExtensions.cs#L43) we call the PublishProgress making its BytesReceived equal 0 and its TotalBytesToReceive the size of the file. But since I'm updating ByteTransferProgress  with your suggestion, I'll have to submit a successful notification explicitly. So, I'll update this.